### PR TITLE
Add UpdatePartitionSpec operation for tables.

### PR DESCRIPTION
**Pull request by rdblue**
This adds `UpdatePartitionSpec` to ` Table` and `Transaction`.

These changes are extracted from #922 and @jun-he is the primary author.

### CODE REVIEW

**Comment by rdblue** [comment]
@jun-he, I've extracted the partition spec update operation from #922 and added it on top of #1919. Most of these changes are yours, except I swapped out the API. There are also some additional changes in #922 that I didn't add because I wanted to keep this small. It would be great to have some of those refactors done, like updating tests to use the new API, if you are interested in contributing those changes in separate PRs or rebasing #922.

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java [hunk 1]
```diff
@@ -501,7 +501,7 @@ public void testNewTableMetadataReassignmentAllIds() throws Exception {
 
     PartitionSpec spec = PartitionSpec.builderFor(schema).withSpecId(5)
         .add(3, 1005, "x_partition", "bucket[4]")
-        .add(5, 1005, "z_partition", "bucket[8]")
+        .add(5, 1003, "z_partition", "bucket[8]")
```
Comment (aokolnychyi) [inline core/src/test/java/org/apache/iceberg/TestTableMetadata.java:504]: This was a typo before?

**Review by aokolnychyi — APPROVED** [review]
+1

File: core/src/test/java/org/apache/iceberg/TestTableMetadata.java (hunk 1 above)
Comment (rdblue) [reply core/src/test/java/org/apache/iceberg/TestTableMetadata.java:504]: The intent is to check whether the IDs are reassigned. The spec builder may reject duplicate IDs.

**Comment by rdblue** [comment]
Thanks for reviewing, @aokolnychyi!

**Comment by rdblue** [comment]
Thanks for reviewing, @aokolnychyi!

**Pull request merged**